### ISSUE TITLE
data storage option

### ISSUE THREAD

**Issue by yuuuxt** [man-group/dtale#236]
First of all thanks for creating this awesome project and continuously improving it!

I'm thinking of using `dtale` in the following way:
1. create a simple backend using `flask` or `fastapi` to simply load/unload `DataFrame` into memory along with some other queries that does not mutate the `DataFrame` (the "other queries" could be some business logic checks that I'd like to perform on the data);
2. let `dtale` uses the loaded `DataFrame` directly without creating a copy.

Currently, it seems that using `dtale.show(df)` will create a copy of the `DataFrame`, doubling the memory usage (also taking considerable time). Would this be possible:
1. have `dtale` started in advance when I start my own simple backend; 
2. set the data storage to what I created in that simple backend;
3. then when I want to check a `DataFrame` in `dtale`, no copy would be made and it would take no extra time to load.

Could you give me some instructions on how to do it? (or would it be supported in the future?)

**Comment by yuuuxt** [comment]
Hi Andrew, it's just a kindly reminder to ensure you didn't miss this issue for seeing you reply to later issues. Look forward to your comment :) @aschonfeld

**Comment by aschonfeld** [comment]
@yuuuxt Sorry, thanks for submitting this.  One of the biggest drawbacks of D-Tale is that it does effectively double memory by making a copy of the dataframe you've passed into it.  The reason for the copy is because up front it calls `df.reset_index()` on the dataframe so that it no longer has to manage index information.

So one thing that I can do is add a flag that will allow this call to `reset_index()` to be `inplace=True`.  I'm thinking that should do the trick, but at the same time it will be altering the state of your variable.  If that doesn't bother you, then it should be a huge improvement.

Here is the [line of code](https://github.com/man-group/dtale/blob/e5f07dbfb39b6380a0b8c3bcc1c6d40d6144ce60/dtale/views.py#L633) in question where the `reset_index()` is being called.

**Comment by yuuuxt** [comment]
> 
> 
> @yuuuxt Sorry for the wait, thanks for submitting this. One of the biggest drawbacks of D-Tale is that it does effectively double memory by making a copy of the dataframe you've passed into it. The reason for the copy is because up front it calls `df.reset_index()` on the dataframe so that it no longer has to manage index information.
> 
> So one thing that I can do is add a flag that will allow this call to `reset_index()` to be `inplace=True`. I'm thinking that should do the trick, but at the same time it will be altering the state of your variable. If that doesn't bother you, then it should be a huge improvement.
> 
> Here is the [line of code](https://github.com/man-group/dtale/blob/e5f07dbfb39b6380a0b8c3bcc1c6d40d6144ce60/dtale/views.py#L633) in question where the `reset_index()` is being called.

For those like me who load data using `pd.read_csv` without setting `index_col`, this no-copy behavior would be more convenient. 

Furthermore, maybe having `inplace=True` and `drop=True` set at the same time is better, under which circumstance calling `reset_index` doesn't make any change to the `DataFrame` (that doesn't add a new column, which also saves a little execution time). 

If that makes sense, could you think of implementing it? 

Besides, I'm also trying to find a way to start dtale beforehand (without passing any data to it) in my own backhand. This is not a big thing, but if there's already a way to do that, I'm happy to learn it.

**Comment by aschonfeld** [comment]
@yuuuxt sorry, I didn't know you were loading from csv files. In that case you do have a couple options available already:
1) You can specify a loading function so that memory hasn't been allocated to a variable beforehand:
```
def _loader():
    return pd.read_csv('test.csv')
dtale.show(data_loader=_loader)
```

2) You can use the pre-assembled CSV loader: `dtale.show_csv('test.csv')`
  * here's some more documentation on the [`show_csv` function](https://github.com/man-group/dtale#accessing-cli-loaders-in-notebook-or-console)

The key to both these options are that they are not storing your dataframe in a variable before calling `dtale.show` so it occupies less memory.  The reason why I end up making a copy when calling `dtale.show` directly is in the case that users want to use D-Tale while debugging code and don't want to accidentally alter the state of their dataframes.  I will be working on adding an `inplace=True` parameter for `dtale.show` so that you can load a dataframe directly w/o duplicating memory like we discussed, but it will still be a conservative approach.

Hope this helps :)

**Comment by yuuuxt** [comment]
@aschonfeld Thank you for offering those options, I like the loading function, and will definitely use it in the future! (I may also load parquet data, so the second option may be limited).

Meanwhile, I will be expecting the direct-load way, as it fits my expected work flow better.  I may want to have my own checks before/after using dtale, so avoiding copying does mean a lot to me.  

Also, I mentioned the `drop=True` option, because it will help save time when I don't care about index. I tested the execution time on a large `DataFrame` (tens of millions of lines), and I found that `drop=True` would save  noticeable time. Hope you would consider adding it too :)

**Comment by aschonfeld** [comment]
@yuuuxt I'll definitely add those two options and if the calling `reset_index` only ends up adding an `index` column with values 0 to "n-1" then I'll drop that regardless.  I think that's how D-tale works already anyways.

Also, after hearing your comment about parquet files I think I'll add a `show_parquet` function as well. :)

Will keep you posted on when this is released

**Comment by aschonfeld** [comment]
@yuuuxt I've just released v1.13.0 to pypi and it includes both an `inplace` & `drop_index` parameter.  I've done memory profiling and by setting `inplace=True` it will cut memory in half.  Here's an example:
```
dtale.show(df, inplace=True)
```

You can also drop any index that exists (although by default if there isn't really an index to being with, for example a natural index from 0 to n-1, it will just drop it):
```
dtale.show(df, inplace=True, drop_index=True)
```

Eventually I might have `inplace` set to `True` by default and just stress in the docs that if you're using this while debugging it will alter the state of your variable.

**Comment by yuuuxt** [comment]
@aschonfeld Thank you for implementing it! I will try it out at weekend :)
